src/metabase/search/filter.clj

```diff
@@ -0,0 +1,124 @@
+;; ------------------------------------------------------------------------------------------------;;
+
+(defmulti ^:private archived-clause
+  "Clause to filter by the archived status of the entity."
+  {:arglists '([model archived?])}
+  (fn [model _] model))
+
+(defmethod archived-clause :default
+  [model archived?]
+  [:= (search.config/column-with-model-alias model :archived) archived?])
+
+;; Databases can't be archived
+(defmethod archived-clause "database"
+  [_model archived?]
+  (if archived?
+    false-clause
+    true-clause))
+
+(defmethod archived-clause "indexed-entity"
+  [_model archived?]
+  (if-not archived?
+    true-clause
+    false-clause))
+
+;; Table has an `:active` flag, but no `:archived` flag; never return inactive Tables
+(defmethod archived-clause "table"
+  [model archived?]
+  (if archived?
+    false-clause                        ; No tables should appear in archive searches
+    [:and
+     [:= (search.config/column-with-model-alias model :active) true]
+     [:= (search.config/column-with-model-alias model :visibility_type) nil]]))
+
+;; ------------------------------------------------------------------------------------------------;;
+;;                                         Optional filters                                        ;;
+;; ------------------------------------------------------------------------------------------------;;
+
+(defmulti ^:private created-by-clause
+  "Clause to filter by the creator of the entity."
+  {:arglists '([model creator-id])}
+  (fn [model _creator-id]
+    model))
+
+(defmethod created-by-clause :default
+  [model _creator-id]
+  (throw (ex-info (format "Created by filter for %s is not supported" model) {:model model})))
+
+(defmethod created-by-clause "card"
+  [model creator-id]
+  [:= (search.config/column-with-model-alias model :creator_id) creator-id])
+
+(defmethod created-by-clause "dataset"
+  [model creator-id]
+  [:= (search.config/column-with-model-alias model :creator_id) creator-id])
+
+(defmethod created-by-clause "dashboard"
+  [model creator-id]
+  [:= (search.config/column-with-model-alias model :creator_id) creator-id])
+
+(defmethod created-by-clause "action"
+  [model creator-id]
+  [:= (search.config/column-with-model-alias model :creator_id) creator-id])
+
+;; ------------------------------------------------------------------------------------------------;;
+;;                                        Public functions                                         ;;
+;; ------------------------------------------------------------------------------------------------;;
+
+(mu/defn search-context->applicable-models :- [:set SearchableModel]
+  "Given a search-context, retuns the list of models that can be applied to it.
+
+  If the context has optional filters, the models will be restricted for the set of supported models only."
+  [{:keys [created-by models] :as search-context} :- SearchContext]
+  (if-not (has-optional-filters? search-context)
+    models
```

Comment:
This if branching is unnecessary, and `has-optional-filters?` can be removed. The second branch is sufficient with `cond->`, because if there are no optional filters the result should be just `models` anyway